observer/observer.go

```diff
@@ -0,0 +1,113 @@
+package observer
+
+import (
+	"context"
+	"errors"
+
+	errorsmod "cosmossdk.io/errors"
+	abcitypes "github.com/cometbft/cometbft/abci/types"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cometbft/cometbft/libs/pubsub/query"
+	rpchttp "github.com/cometbft/cometbft/rpc/client/http"
+	coretypes "github.com/cometbft/cometbft/rpc/core/types"
+	comettypes "github.com/cometbft/cometbft/types"
+)
+
+const ModuleNameObserver = "observer"
+
+type Observer struct {
+	logger        log.Logger
+	tmRpc         *rpchttp.HTTP
+	stopChan      chan struct{}
+	eventsOutChan chan abcitypes.Event
+}
+
+// NewObserver returns new instance of `Observer` with RPC client created
+func NewObserver(logger log.Logger, rpcUrl string) (Observer, error) {
+	if len(rpcUrl) == 0 {
+		return Observer{}, errors.New("RPC URL can't be empty")
+	}
+
+	rpc, err := rpchttp.New(rpcUrl, "/websocket")
+	if err != nil {
+		return Observer{}, errorsmod.Wrapf(err, "Failed to create RPC client")
+	}
+
+	return Observer{
+		logger:        logger.With("module", ModuleNameObserver),
+		tmRpc:         rpc,
+		stopChan:      make(chan struct{}),
+		eventsOutChan: make(chan abcitypes.Event),
+	}, nil
+}
+
+// Starts RPC client, subscribes to events for provided query and starts listening to the events
+func (o *Observer) Start(ctx context.Context, queryStr string, observeEvents []string) error {
+	err := o.tmRpc.Start()
+	if err != nil {
+		return err
+	}
+
+	query, err := query.New(queryStr)
+	if err != nil {
+		return errorsmod.Wrapf(err, "Invalid query")
+	}
+
+	txs, err := o.tmRpc.Subscribe(ctx, "observer", query.String())
+	if err != nil {
+		return errorsmod.Wrapf(err, "Failed to subscribe to RPC client")
+	}
+
+	o.logger.Info("Observer starting listening for events at RPC", o.tmRpc.Remote())
+	go o.processEvents(ctx, txs, observeEvents)
+
+	return nil
+}
+
+// Stops listening to events, unsubscribes from RPC client and stops RPC channel
+func (o *Observer) Stop(ctx context.Context) error {
+	close(o.stopChan)
+	if err := o.tmRpc.UnsubscribeAll(ctx, "observer"); err != nil {
+		return errorsmod.Wrapf(err, "Failed to unsubscribe from RPC client")
+	}
+	close(o.eventsOutChan)
```

Comment:
This code will panic. `Stop` and `processEvents` methods are executed in different goroutines. Imagine this line is executed while the other goroutine is only completing executing the last event. In that case we will try to write into the closed channel (`o.eventsOutChan <- e` line 107) and consequently get a panic. Thus we must close this channel strictly after the last write. To do that you can add `defer close(o.eventsOutChan)` at the first line of `processEvents`